java/arcs/core/util/Json.kt

```diff
@@ -0,0 +1,187 @@
+
+/** Visitor interface for [JsonValue] hierarchy. */
+interface JsonVisitor<R> {
+    /** Primary entrypoint to visitor. */
+    fun visit(value: JsonValue<*>) = when(value) {
+        is JsonBoolean -> visit(value as JsonBoolean)
+        is JsonString -> visit(value as JsonString)
+        is JsonNumber -> visit(value as JsonNumber)
+        is JsonArray -> visit(value as JsonArray)
+        is JsonObject -> visit(value as JsonObject)
+        is JsonNull -> visit(value as JsonNull)
+        else -> throw IllegalArgumentException("Unknown JsonValue type $value")
+    }
+
+    /** Called when a [JsonBoolean] is encountered. */
+    fun visit(value: JsonBoolean): R
+
+    /** Called when a [JsonString] is encountered. */
+    fun visit(value: JsonString): R
+
+    /** Called when a [JsonNumber] is encountered. */
+    fun visit(value: JsonNumber): R
+
+    /** Called when a [JsonArray] is encountered. */
+    fun visit(value: JsonArray): R
+
+    /** Called when a [JsonNull] is encountered. */
+    fun visit(value: JsonObject): R
+
+    /** Called when a [JsonNull] is encountered. */
+    fun visit(value: JsonNull): R
+}
+
+sealed class JsonValue<T>() {
+    abstract val value: T
+    /** Return this as a [JsonArray] or null if it isn't an array. */
+    fun array() = this as? JsonArray
+
+    /** Return this as a [JsonObject] or null if it isn't an object. */
+    fun obj() = this as? JsonObject
+
+    /** Return this as a [JsonNumber] or null if it isn't a number. */
+    fun number() = (this as? JsonNumber)?.value
+
+    /** Return this as a [JsonString] or null if it isn't a string. */
+    fun string() = (this as? JsonString)?.value
+
+    /** Return this as a [JsonBoolean] or null if it isn't a boolean. */
+    fun bool() = (this as? JsonBoolean)?.value
+
+    /** Used to represent the value of 'null' */
+    object Null {
+        override fun toString() = "null"
+    }
+
+    /** Used to represent a Json string */
+    data class JsonString(override val value: String) : JsonValue<String>() {
+        override fun toString() = "\"${value.replace("\"", "\\\"")}\""
+    }
+
+    /** Used to represent a Json number */
+    data class JsonNumber(override val value: Double) : JsonValue<Double>() {
+        override fun toString() = value.toString()
+    }
+
+    /** Used to represent a Json boolean */
+    data class JsonBoolean(override val value: Boolean) : JsonValue<Boolean>() {
+        override fun toString() = value.toString()
+    }
+
+    /** Used to represent custom (potentially non-standard) Json values. */
+    open class JsonAny<T>(override val value: T) : JsonValue<T>()
+
+    /** Used to to represent a Json array .*/
+    data class JsonArray(
+        override val value: List<JsonValue<*>> = mutableListOf()
+    ) : JsonValue<List<JsonValue<*>>>() {
+        override fun toString() = value.joinToString(prefix = "[", postfix = "]", separator = ",")
+
+        /** Number of elements in the array. */
+        val size = value.size
```

Comment:
should be a getter?

Reply:
You're right. In the immutable case it works, but not in the mutable one. Good catch.